pallets/inactivity-tracking/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The size of a collator set for a session has already reached MaxCollatorsPerSession value
+        MaxCollatorsPerSessionReached,
+        /// Error returned when the activity tracking status is attempted to be updated before the end session
+        ActivityStatusUpdateSuspended,
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        #[pallet::call_index(0)]
+        #[pallet::weight(T::WeightInfo::set_inactivity_tracking_status())]
+        pub fn set_inactivity_tracking_status(
+            origin: OriginFor<T>,
+            is_enabled: bool,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+            let current_status_end_session_index = match <CurrentActivityTrackingStatus<T>>::get() {
+                ActivityTrackingStatus::Enabled { start: _, end } => end,
+                ActivityTrackingStatus::Disabled { end } => end,
+            };
+            let current_session_index = T::CurrentSessionIndex::session_index();
+            ensure!(
+                current_session_index > current_status_end_session_index,
+                Error::<T>::ActivityStatusUpdateSuspended
+            );
+            let new_status_end_session_index =
+                current_session_index + T::MaxInactiveSessions::get();
+            let new_status = if is_enabled {
+                ActivityTrackingStatus::Enabled {
+                    start: current_session_index + 1,
+                    end: new_status_end_session_index,
+                }
+            } else {
+                <ActiveCollatorsForCurrentSession<T>>::put(BoundedBTreeSet::new());
+                ActivityTrackingStatus::Disabled {
+                    end: new_status_end_session_index,
+                }
+            };
+            <CurrentActivityTrackingStatus<T>>::put(new_status.clone());
+            Self::deposit_event(Event::<T>::ActivityTrackingStatusSet { status: new_status });
+            Ok(())
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+            let mut total_weight = T::DbWeight::get().reads_writes(1, 0);
+
+            // Process the orchestrator chain block author (if it exists) and activity tracking is enabled
+            if let Some(orchestrator_chain_author) = T::GetSelfChainBlockAuthor::get_block_author()
+            {
+                if let ActivityTrackingStatus::Enabled { start, end: _ } =
+                    <CurrentActivityTrackingStatus<T>>::get()
+                {
+                    if start <= T::CurrentSessionIndex::session_index() {
+                        total_weight += Self::on_author_noted(orchestrator_chain_author);
+                    }
+                }
+            }
+            total_weight
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn process_ended_session() {
```

Comment:
I still think we should separate this into:

- things that should execute when session ends through the trait (store the current active collators).
- things that should execute when the session starts (put an empty btree map)

This would avoid having that saturating.sub which is rather ugly. But we can leave this for another PR/issue. can you create one for this?¿

Reply:
Yeah fair point. I extended `pallet-initializer`'s SessionHandler trait in https://github.com/moondance-labs/tanssi/pull/899 to also execute `on_before_session_ending`. Perhaps can updating the `ActiveCollators` for the session in that PR too?

Reply:
let's leave it for another PR, just create an issue to move it there, I dont want to make this bigger